mm2src/coins/eth/eth_balance_events.rs

```diff
@@ -0,0 +1,161 @@
+              stream::FuturesUnordered,
+              StreamExt};
+use instant::Instant;
+use mm2_core::mm_ctx::MmArc;
+use mm2_err_handle::prelude::MmError;
+use mm2_event_stream::{behaviour::{EventBehaviour, EventInitStatus},
+                       Event, EventStreamConfiguration};
+use mm2_number::BigDecimal;
+use std::collections::BTreeMap;
+
+use super::EthCoin;
+use crate::{eth::{u256_to_big_decimal, Erc20TokenInfo},
+            BalanceError, MmCoin};
+
+/// This implementation differs from others, as they immediately return
+/// an error if any of the requests fails. This one completes all futures
+/// and returns their results individually.
+async fn get_all_balance_results_concurrently(
+    coin: &EthCoin,
+) -> Vec<Result<(String, BigDecimal), (String, MmError<BalanceError>)>> {
+    let mut tokens = coin.get_erc_tokens_infos();
+
+    // Workaround for performance purposes.
+    //
+    // Unlike tokens, the platform coin length is constant (=1). Instead of creating a generic
+    // type and mapping the platform coin and the entire token list (which can grow at any time), we map
+    // the platform coin to Erc20TokenInfo so that we can use the token list right away without
+    // additional mapping.
+    tokens.insert(coin.ticker.clone(), Erc20TokenInfo {
+        token_address: coin.my_address,
+        decimals: coin.decimals,
+    });
+
+    let jobs = tokens
+        .into_iter()
+        .map(|(token_ticker, info)| async move {
+            if token_ticker == coin.ticker {
+                let balance_as_u256 = coin
+                    .address_balance(coin.my_address)
+                    .compat()
+                    .await
+                    .map_err(|e| (token_ticker.clone(), e))?;
+                let balance_as_big_decimal = u256_to_big_decimal(balance_as_u256, coin.decimals)
+                    .map_err(|e| (token_ticker.clone(), e.into()))?;
+                Ok((coin.ticker.clone(), balance_as_big_decimal))
+            } else {
+                let balance_as_u256 = coin
+                    .get_token_balance_by_address(info.token_address)
+                    .await
+                    .map_err(|e| (token_ticker.clone(), e))?;
+                let balance_as_big_decimal = u256_to_big_decimal(balance_as_u256, info.decimals)
+                    .map_err(|e| (token_ticker.clone(), e.into()))?;
+                Ok((token_ticker, balance_as_big_decimal))
+            }
+        })
+        .collect::<FuturesUnordered<_>>();
+
+    jobs.collect().await
+}
+
+#[async_trait]
+impl EventBehaviour for EthCoin {
+    const EVENT_NAME: &'static str = "COIN_BALANCE";
+    const ERROR_EVENT_NAME: &'static str = "COIN_BALANCE_ERROR";
+
+    async fn handle(self, interval: f64, tx: oneshot::Sender<EventInitStatus>) {
+        const RECEIVER_DROPPED_MSG: &str = "Receiver is dropped, which should never happen.";
+
+        async fn with_socket(_coin: EthCoin, _ctx: MmArc) { todo!() }
```

Comment:
This is blocked by https://github.com/KomodoPlatform/atomicdex-gui-auth/pull/18 implementation. And not a blocker in my opinion (if we review/merge this PR at this stage, GUI devs can start implementing balance streaming for eth,cosmos,utxo protocols).